### ISSUE TITLE
Move to https by default

### ISSUE THREAD

**Issue by tomhughes** [openstreetmap/operations#190]
The majority of our services are already forcing https by redirecting any http requests. I'd like to move the remaining services to do the same. I believe that to be:

* www.openstreetmap.org
* tile.openstreetmap.org
* gps-tile.openstreetmap.org
* nominatim.openstreetmap.org
* planet.openstreetmap.org
* wiki.openstreetmap.org
* dev.openstreetmap.org
* os.openstreetmap.org
* ooc.openstreetmap.org
* agri.openstreetmap.org

There are some technical issues with the main web site that I need to work on but the rest is easy enough I believe, unless there are concerns about load.

**Comment by tomhughes** [comment]
Quick attempt at a check list of things to do:

* www.openstreetmap.org
  - [ ] figure out a way to resolve oauth issues
  - [ ] add redirect
* tile.openstreetmap.org
  - [ ] make sure render servers have suitable certificates
  - [ ] upgrade backhaul to use https
  - [ ] update leaflet-osm etc to always use https
  - [ ] add redirect (assuming we're happy with potential query increase)
* nominatim.openstreetmap.org
  - [ ] check if @lonvia has any worries around load
  - [ ] add redirect
* wiki.openstreetmap.org
  - [ ] check if @Firefishy knows of any reason we're not already doing it
  - [ ] add redirect

Once all that is in place we can think about deploying HSTS across everything.

**Comment by Firefishy** [comment]
I'm not aware of any reason why we shouldn't enabled HTTPS by default on wiki.openstreetmap.org (with a redirect). There is a likelihood that there will be some 3rd parties who many need to enable HTTPS support in their code eg: taginfo.

**Comment by tomhughes** [comment]
Taginfo already redirects everything to https anyway.

**Comment by lonvia** [comment]
I have no big worries about load for nominatim.openstreetmap.org.

I suspect though that for nominatim and planet there are quite a few scripts in active use which cannot handle redirects. Maybe we should consider some kind of pre-announcement.

**Comment by tomhughes** [comment]
Really? Are there many http clients that don't handle redirects?

I guess main shell scripts using curl, which for reasons I've never understood doesn't do redirects unless you add the `-L` option.

**Comment by SomeoneElseOSM** [comment]
> Taginfo already redirects everything to https anyway.
It does, but not all the country ones necessarily do (taginfo.openstreetmap.org.uk for example doesn't seem to have a valid certificate on it).  Obviously that's a separate issue (and I'll mention it to the maintainer of that site), and a fairly low-level issue since it's not the main one.

**Comment by tomhughes** [comment]
Well whatever - wiki is already available over https and lots of people use it that way so it can't exactly have any major problems.

**Comment by tomhughes** [comment]
I now have that is (hopefully) a [workaround](https://github.com/tomhughes/openstreetmap-website/commit/b6e83ccab8093e4f8fcd31b4d968564d9dd3679e) the the OAuth issues on the main site and I have that running at https://tomh.apis.dev.openstreetmap.org/ if anybody wants to test OAuth against that (using an http URL that will redirect).

**Comment by pnorman** [comment]
Testing on planet.openstreetmap.org revealed that osmosis doesn't follow redirects: https://trac.openstreetmap.org/ticket/5483

This can be worked around by switching the `http://planet.openstreetmap.org` to `https://planet.openstreetmap.org` in the osmosis `configuration.txt` file, but users need time to make the change.

I'm writing an email to inform people.

**Comment by genodeftest** [comment]
Adding HSTS to the wiki would be nice. `www.` and `dev.` have it already. I don't know about the others.

**Comment by tomhughes** [comment]
It should have it already.

**Comment by genodeftest** [comment]
@tomhughes wrote:
> It should have it already.

I don't see it.

**Comment by tomhughes** [comment]
https://github.com/openstreetmap/chef/commit/5239c5f1a1d13c8c34024e01f8c98282fa085181 should have fixed it.

**Comment by genodeftest** [comment]
@tomhughes: 
Now I see it on the wiki too. Thanks!

**Comment by mmd-osm** [comment]
Could we also add https://wiki.openstreetmap.org/wiki/Slippy_Map_MediaWiki_Extension to the todo list? It's still using http to fetch tiles, including mixed content warnings. Those seem to be hard coded here: https://www.openstreetmap.org/openlayers/OpenStreetMap.js

// ping: @harry-wood

**Comment by tomhughes** [comment]
The only that extension needs is banishing to the pits of doom.

I have absolutely no plans to change that archived copy of OpenLayers which nobody (including that extension!) should be using anyway - we only keep it around because so many people in the early days decided it was a good idea to load OpenLayers from our site.

**Comment by tomhughes** [comment]
Hmm are we actually the canonical source for that file in fact? or is that part of upstream OpenLayers?

Still changing it is scary, and breaking all the sites loading OL from us does have a certain appeal if it makes some of them upgrade ;-)

**Comment by tomhughes** [comment]
Looks like that file is actually our code - there is a generic OSM layer in upstream and it creates layer specific classes derived from that.

In any cases the images are passive content so won't currently be a problem - the big problem is that the js is active content so won't load over http from an https context. That will need fixing in the extension itself.

**Comment by tomhughes** [comment]
Hmm those loads are protocol relative. That's weird then...

**Comment by tomhughes** [comment]
It's actually bombing out creating an instance of `OpenLayers.Control.Button` because there is no such class in that OL we are serving. That's why you just get pink tiles.

**Comment by mmd-osm** [comment]
@tomhughes : Thanks! Switch to https looks good here, Javascript warnings no longer show up in the console. I will also check with the user who originally reported the issue on the Forum, if the tiles are now visible in IE11, or if there's still some other issue.

**Comment by grischard** [comment]
Quick update:

- tile.openstreetmap.org
  [x] update leaflet-osm etc to always use https

- planet.openstreetmap.org
  [x] persuade users to switch to using https (osmium doesn't follow redirects)
  [x] add redirect

**Comment by Firefishy** [comment]
@grischard The HSTS spec makes clear clients MUST NOT trust HSTS header from HTTP sources. I don't think it causes any harm if we public and clients ignore it.

**Comment by grischard** [comment]
@Firefishy yes, and many follow them anyway, which makes it a neat soft redirect hack. There's a discussion about it at https://github.com/openstreetmap/operations/issues/117#issuecomment-259192002

**Comment by grischard** [comment]
I see that wiki.openstreetmap.org and plain openstreetmap.org currently don't set the HSTS header. Should we enable it everywhere, add `includeSubDomains`, and add `preload` eventually if we're happy?

**Comment by tomhughes** [comment]
Both those domains seem to set it for me:

```
bericote [~] % HEAD -P https://wiki.openstreetmap.org/ | fgrep Strict-Transport-Security 
Strict-Transport-Security: max-age=31536000
bericote [~] % HEAD -P https://openstreetmap.org/ | fgrep Strict-Transport-Security     
Strict-Transport-Security: max-age=31536000
```
I think `hot.openstreetmap.org` is the main outstanding issue for preload?

**Comment by grischard** [comment]
Ah, I see, it's set on `https://wiki.openstreetmap.org/wiki/Main_Page` but not on `https://wiki.openstreetmap.org` that redirects to it, and it's set on `https://www.openstreetmap.org` but not on `https://openstreetmap.org` that redirects to it. I don't know if that's moot or not for now, but we'll need it on https://openstreetmap.org the day we add `preload`, I think.

Who's running `hot.openstreetmap.org` again? The *.hotosm.org certificate expires in two months, so now would be a good time for them to work on this.

**Comment by grischard** [comment]
CC'ing @smit1678 - would it please be possible to add a letsencrypt certificate for [hot, humanitarian, hot-dev and hot-staging.openstreetmap.org](https://trac.openstreetmap.org/browser/dns/src/openstreetmap#L481) on amelia.hotosm.org? Alternatively, if only hot.openstreetmap.org is being used, and it's being used only for a redirect, we can bring hot.openstreetmap.org back to the osm infrastructure, have a certificate there, and just redirect to hotosm.org from there.

**Comment by bgirardot** [comment]
Hi all,

Stereo contacted HOT via IRC and we now have a ticket for getting this supported on our end.

I have created a ticket for it in our system as well, which was auto linked above in this issue.

@smith1678 or HOT's sys admin @dakotabenjamin will follow up.

cheers

**Comment by smit1678** [comment]
> Alternatively, if only hot.openstreetmap.org is being used, and it's being used only for a redirect, we can bring hot.openstreetmap.org back to the osm infrastructure, have a certificate there, and just redirect to hotosm.org from there.

@grischard Thanks for flagging this. It looks like it's also redirecting `hot.osm.org` as well. I think it's probably best if we do this on the OSM infrastructure since this is a `openstreetmap.org` subdomain and just a redirect. Having it continue to redirect to hotosm.org would be great. We've been slowly transitioning some things off of the server that it's currently pointing to and so I think for the long term it would be best for that subdomain managed with everything else for osm.org. 

If that's not the case, we can certainly set up the certificate to do the redirect correctly.

**Comment by grischard** [comment]
@tomhughes would hosting the hot.o.o redirect ourselves be ok?

Also, I don't know if you saw my message on IRC: I'm thinking of doing a PR for a full squid setup for the tiles, it should be enough for now, and less work than varnish while still making an eventual future switch possible. Good idea?

**Comment by smit1678** [comment]
@tomhughes Yes, that's correct, we don't use those either. Amelia is the nickname of the server and that redirect was implemented a while ago.

**Comment by tomhughes** [comment]
I have moved hot.openstreetmap.org to point at one of our machines that has a valid certificate and just redirects to www.hotosm.org now.

**Comment by tomhughes** [comment]
@grischard by "full squid" you mean dropping nginx and using squid to terminate https connections as well as http? that sounds good to me...

**Comment by pnorman** [comment]
tile has moved to redirects from http to https, so all the items in the list are done. We can consider HSTS in a different issue

**Comment by tomhughes** [comment]
We've been HSTS preloaded for some time...

**Comment by grischard** [comment]
Please re-open, tile is indeed the last one that still doesn't redirect.

How to reproduce:

`curl -v a.tile.openstreetmap.org/10/11/12.png`

Expected:

redirect

Actual result:

200 OK and PNG

**Comment by grischard** [comment]
The tiles are a hairy yak to shave. The rate limiting stuff we use in squid only works in that version of squid, which doesn't really support modern https and http/2. Upgrading squid breaks things. Re-writing this completely in nginx is possible but would take time, and Andy says that this kind of pre-rendered tile architecture is on its way out in the medium term anyway.
